pandas/_libs/src/vendored/ujson/python/objToJSON.c

```diff
@@ -1538,13 +1550,24 @@ void Object_beginTypeContext(JSOBJ _obj, JSONTypeContext *tc) {
         tc->type = JT_UTF8;
         return;
     } else if (PyArray_IsScalar(obj, Datetime)) {
+        npy_int64 longVal;
         if (((PyDatetimeScalarObject *)obj)->obval == get_nat()) {
             tc->type = JT_NULL;
             return;
         }
+        PyArray_Descr *dtype = PyArray_DescrFromScalar(obj);
+        if (dtype->type_num == NPY_OBJECT) {
```

Comment:
Wondering if this can be removed. This shouldn't ever fail, but if it does, we'll end up with a nasty segfault.

Reply:
https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_DescrFromScalar

Reading through that makes me think we are supposed to check we are dealing with an array scalar before even calling this function,

As far as the resulting comparison to NPY_OBJECT, I think would be better to explicitly use the `PyTypeNum_ISDATETIME` macro and error when that is not true; that would be a better and more explicit safeguard against troublesome refactors

Reply:
Yeah, this whole block is in a PyArray_IsScalar if block.